### PR TITLE
[sources] prevent unnecessary parsing

### PR DESCRIPTION
**Pull request by jasonLaster**
### Summary of Changes

This is a really nice performance win I saw by looking at perf.html and I thought that I'd get the win while i could. Last week we switched to memoizing getURL, this opened up the opportunity to prevent doing a ton of unnecessary parsing:

- debuggeeHost 
- file extensions
- relative urls...

screen shot 2018-07-06 at 2 55 40 pm

### CODE REVIEW

File: src/selectors/breakpointSources.js [hunk 1]
```diff
@@ -53,19 +53,12 @@ function findBreakpointSources(
...
-
 export const getBreakpointSources = createSelector(
   getBreakpoints,
   getSources,
-  _getBreakpointSources
+  (breakpoints: BreakpointsMap, sources: SourcesMap) =>
+    findBreakpointSources(sources, breakpoints).map(source => ({
+      source,
+      breakpoints: getBreakpointsForSource(source, breakpoints)
+    }))
```
Comment (jasonLaster) [inline src/selectors/breakpointSources.js]: just simplifying...

File: src/selectors/getRelativeSources.js [hunk 2]
```diff
@@ -7,20 +7,22 @@
...
-import { getSourcePath } from "../utils/source";
+import { getURL } from "../utils/sources-tree";
 import { createSelector } from "reselect";
 
-function getRelativeUrl(url, root) {
+function getRelativeUrl(source, root) {
+  const { group, path } = getURL(source);
   if (!root) {
-    return getSourcePath(url);
+    return path;
```
Comment (jasonLaster) [inline src/selectors/getRelativeSources.js]: re-computing the source path can be expensive. It is better to use `getURL` which is cached

File: src/utils/sources-tree/addToTree.js [hunk 3]
```diff
@@ -161,11 +161,10 @@ function addSourceToNode(
 export function addToTree(
   tree: TreeDirectory,
   source: Source,
-  debuggeeUrl: string,
+  debuggeeHost: ?string,
   projectRoot: string
 ) {
-  const url = getURL(source, debuggeeUrl);
-  const debuggeeHost = getDomain(debuggeeUrl);
+  const url = getURL(source, debuggeeHost);
```
Comment (jasonLaster) [inline src/utils/sources-tree/addToTree.js]: lets pass `debuggeeHost` in so we don't recompute it each time

File: src/utils/sources-tree/utils.js [hunk 4]
```diff
@@ -49,16 +50,16 @@ export function isSource(item: TreeNode) {
   return item.type === "source";
 }
 
-export function getFileExtension(url: string = ""): string {
-  const parsedUrl = parse(url).pathname;
+export function getFileExtension(source: Source): string {
+  const parsedUrl = getURL(source).path;
```
Comment (jasonLaster) [inline src/utils/sources-tree/utils.js]: lets use `getURL` which is cached

**Comment by darkwing** [comment]
Working on this now; also, after setting a root and refreshing, I see this error:

```js
The above error occurred in the <SourceIcon> component:
    in SourceIcon (created by Connect(SourceIcon))
    in Connect(SourceIcon) (created by SourceTreeItem)
    in div (created by SourceTreeItem)
    in SourceTreeItem (created by TreeNode)
    in div (created by TreeNode)
    in TreeNode
    in div (created by Tree)
    in Tree (created by ManagedTree)
    in div (created by ManagedTree)
    in ManagedTree (created by SourcesTree)
    in div (created by SourcesTree)
    in div (created by SourcesTree)
    in SourcesTree (created by Connect(SourcesTree))
    in Connect(SourcesTree) (created by PrimaryPanes)
    in div (created by PrimaryPanes)
    in PrimaryPanes (created by Connect(PrimaryPanes))
    in Connect(PrimaryPanes) (created by App)
    in div (created by SplitBox)
    in div (created by SplitBox)
    in SplitBox (created by App)
    in div (created by SplitBox)
    in div (created by SplitBox)
    in SplitBox (created by App)
    in div (created by App)
    in App (created by Connect(App))
    in Connect(App)
    in Provider

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://fb.me/react-error-boundaries to learn more about error boundaries.
react-dom.development.js:9747
TypeError: WeakMap key must be an object, got the string "https://firefox-dev.tools/debugger-examples/examples/todomvc/js/views/app-view.js"
getURL.js:126:2

```

File: src/utils/sources-tree/getURL.js [hunk 5]
```diff
@@ -119,12 +117,12 @@ function _getURL(source: Source, debuggeeUrl: string): ParsedURL {
...
 }
 
-export function getURL(source: Source, debuggeeUrl: string = "") {
+export function getURL(source: Source, debuggeeUrl: ?string = "") {
   if (urlMap.has(source)) {
     return urlMap.get(source) || def;
   }
 
-  const url = _getURL(source, debuggeeUrl);
+  const url = _getURL(source, debuggeeUrl || "");
```
Comment (jasonLaster) [inline src/utils/sources-tree/getURL.js]: did `getURL(source: Source, debuggeeUrl?: string = "")` fail?

Comment (jasonLaster) [reply src/utils/sources-tree/getURL.js]: we should be able to avoid this: `debuggeeUrl || ""` if we have the right type

**Pull request merged**